### ISSUE TITLE
lean compile issues

### ISSUE THREAD

**Issue by EgbertRijke** [leanprover/lean3#812]
I had some issues with compiling the hott library. My goal was to compile myself an up to date version of lean (with the help of Floris), so I cloned the code from github. However, when I run make, the hott library doesn't compile. A typical error message that appears in the output is:

`/home/egbert/Documents/Lean/hott/init/num.hlean:22:30: error: unknown identifier 'succ'`

Floris and I just found out that apparently, there is exists an old installation on my Ubuntu system, in `/usr/local/lib`, because in the file `succ.d` we find reference to these. When I go to this folder to inspect the lean files there, I find that these are created in January, maybe when I installed lean from the repository.

So I'd like to remove this installation. However, I cannot remove this installation with `sudo apt-get remove lean`, because it gives the error that it is unable to locate lean. We both don't really know how to proceed.

**Comment by soonhokong** [comment]
@EgbertRijke, I'll see how to manually remove the previously installed lean files.

Apart from that, I think there is another issue. You (1) checkout the source code of Lean, (2) build a Lean binary, and 3) compile standard/hott libraries. In step (3), it should use the lean binary generated in step (2). Maybe that's not the case for now, and it uses the one in the system installed via apt-get. I'll double-check this one as well.

**Comment by fpvandoorn** [comment]
I think the problem is with the generation of the dependency files. For example `num.d`, located in 

```
/home/egbert/Documents/Lean/hott/init/num.d
```

referenced

```
/usr/lib/lean/hott/init/logic.olean
```

(or something similar), instead of

```
/home/egbert/Documents/Lean/hott/init/logic.olean
```

(so `/home/egbert/Documents/Lean` is the location of the repository he just cloned and `/usr/lib/lean/` is the location of the old Lean which wasn't changed since January).

Also: the standard library compiled just fine.

**Comment by soonhokong** [comment]
@EgbertRijke, @fpvandoorn, could you test the following?

``` bash
cd /home/egbert/Documents/Lean/hott
git clean -fxd
../bin/linja
```

**Comment by EgbertRijke** [comment]
I get some warnings:

```
[95/98] "/home/egbert/Documents/Lean/b.../hott/algebra/category/adjoint.ilean" 
/home/egbert/Documents/Lean/hott/algebra/category/adjoint.hlean:138:12: warning: using 'sorry'
/home/egbert/Documents/Lean/hott/algebra/category/adjoint.hlean:164:46: warning: using 'sorry'
/home/egbert/Documents/Lean/hott/algebra/category/adjoint.hlean:165:12: warning: using 'sorry'
```

but otherwise it runs without errors. This just compiled the hott library?

**Comment by EgbertRijke** [comment]
I have all the olean files in the right directory now, and the references in the *.d files seem to be correct. So that solves the issue, I think :) Thanks!

**Comment by soonhokong** [comment]
@EgbertRijke, it seems everything is OK now.

I'm wondering how you got the problem. Did you get the problem when you build Lean by running either `make` or `ninja`? If so, the problem is still there and I want to fix it.

If you got the problem by running `linja` in `Lean/hott` directory, then you got an expected behavior and I don't think there is anything to fix..

**Comment by fpvandoorn** [comment]
We tried all those things. First we tried `ninja`, which didn't work, but we also used `make` and `linja` in `Lean/hott` and removed temporary files using `git clean -Xf`.

But I don't know the exact order of commands anymore. It is perfectly possible that we used `git clean -Xf` in `Lean/hott` and then immediately executed `linja`, and only after that tried `make` again. So I think there is no problem, except an unlucky sequence of commands.
A couple days ago we managed to compile the HoTT library correctly, so then we probably executed `git clean -Xf` followed by `make`.

**Comment by EgbertRijke** [comment]
Now I'm trying to compile Floris' leansnippets repository, and it gives a similar error when executing `linja`:

```
[6/22] "/usr/local/bin/lean"  "/home/e...uments/leansnippets/crazy_hits.ilean" 
FAILED: "/usr/local/bin/lean"  "/home/egbert/Documents/leansnippets/crazy_hits.hlean" 
-o "/home/egbert/Documents/leansnippets/crazy_hits.olean" 
-c "/home/egbert/Documents/leansnippets/crazy_hits.clean" 
-i "/home/egbert/Documents/leansnippets/crazy_hits.ilean"
```

When I inspect any `*.d` file that is generated by linja, it refers to `usr/local/lib/lean`. For example, my file `Wsusp.d` reads:

```
/home/egbert/Documents/leansnippets/Wsusp.olean: \
 /home/egbert/Documents/leansnippets/error: file 'hit/quotient' not found in the LEAN_PATH \
 /usr/local/lib/lean/hott/types/W.olean \
 /usr/local/lib/lean/hott/init/default.olean
```

So `linja` is still referring to the old installation, but `apt-get` doesn't know there exists an old installation. So I'd like that this issue gets reopened, so that we can figure out what is going on.

**Comment by EgbertRijke** [comment]
When I inspect the path of lean by `lean --path`, it gives `/home/egbert/Documents/Lean`, which is the correct installation of lean. So the problem is not in the path of lean. It must be something in linja.

**Comment by soonhokong** [comment]
``` bash
[6/22] "/usr/local/bin/lean"  "/home/e...uments/leansnippets/crazy_hits.ilean" 
```

I can see that `/usr/local/bin/lean` is being used here. What did you execute when you compile Floris' leansnippets repository?

**Comment by EgbertRijke** [comment]
I first did a cleanup with `git clean -xfd` and then I ran `linja`

**Comment by soonhokong** [comment]
OK. What do you have when you run `which linja`?

**Comment by EgbertRijke** [comment]
Ah! it says `/usr/local/bin/linja`

**Comment by soonhokong** [comment]
I see. You need to either explicitly call `linja` in your home directory, or adjust `PATH` variable so that `which linja` points to the one that you want to call.

**Comment by EgbertRijke** [comment]
Which line should I add to `.bashrc` to adjust the path correctly?

**Comment by soonhokong** [comment]
Add `PATH=<path/to/lean/bin>:$PATH` at the end of `~/.bashrc`. This link should help as well: http://unix.stackexchange.com/a/26059

**Comment by EgbertRijke** [comment]
Okay, I managed to do that. It seems to have compiled now, though it run into some errors due to incomplete definitions. But those are due to Floris :P 

Thanks a lot!

**Comment by soonhokong** [comment]
Great! I'm closing this issue.
